# Heads up!

## This notebook is wrong. I left it in because you told us to keep our faulty code. The correct code is in v2. You can look through this if you want to laugh at all the time I wasted making an invalid model

# Project 2 Modeling

### EDA

In [1]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression, LassoCV
from sklearn.preprocessing import StandardScaler, PolynomialFeatures

In [2]:
ames_csv = "./train.csv"
ames = pd.read_csv(ames_csv)
test_csv = "./test.csv"
test_set = pd.read_csv(test_csv)

In [3]:
ames.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley  \
0  109  533352170           60        RL           NaN     13517   Pave   NaN   
1  544  531379050           60        RL          43.0     11492   Pave   NaN   
2  153  535304180           20        RL          68.0      7922   Pave   NaN   
3  318  916386060           60        RL          73.0      9802   Pave   NaN   
4  255  906425045           50        RL          82.0     14235   Pave   NaN   

  Lot Shape Land Contour    ...     Screen Porch Pool Area Pool QC Fence  \
0       IR1          Lvl    ...                0         0     NaN   NaN   
1       IR1          Lvl    ...                0         0     NaN   NaN   
2       Reg          Lvl    ...                0         0     NaN   NaN   
3       Reg          Lvl    ...                0         0     NaN   NaN   
4       IR1          Lvl    ...                0         0     NaN   NaN   

  Misc Feature Misc Val Mo Sold Yr Sold  Sale Type  SalePrice  
0          NaN        0       3    2010        WD      130500  
1          NaN        0       4    2009        WD      220000  
2          NaN        0       1    2010        WD      109000  
3          NaN        0       4    2010        WD      174000  
4          NaN        0       3    2010        WD      138500  

[5 rows x 81 columns]

In [4]:
ames.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2051 entries, 0 to 2050
Data columns (total 81 columns):
Id                 2051 non-null int64
PID                2051 non-null int64
MS SubClass        2051 non-null int64
MS Zoning          2051 non-null object
Lot Frontage       1721 non-null float64
Lot Area           2051 non-null int64
Street             2051 non-null object
Alley              140 non-null object
Lot Shape          2051 non-null object
Land Contour       2051 non-null object
Utilities          2051 non-null object
Lot Config         2051 non-null object
Land Slope         2051 non-null object
Neighborhood       2051 non-null object
Condition 1        2051 non-null object
Condition 2        2051 non-null object
Bldg Type          2051 non-null object
House Style        2051 non-null object
Overall Qual       2051 non-null int64
Overall Cond       2051 non-null int64
Year Built         2051 non-null int64
Year Remod/Add     2051 non-null int64
Roof Style         20

In [5]:
ames.describe()

Id           PID  MS SubClass  Lot Frontage       Lot Area  \
count  2051.000000  2.051000e+03  2051.000000   1721.000000    2051.000000   
mean   1474.033642  7.135900e+08    57.008776     69.055200   10065.208191   
std     843.980841  1.886918e+08    42.824223     23.260653    6742.488909   
min       1.000000  5.263011e+08    20.000000     21.000000    1300.000000   
25%     753.500000  5.284581e+08    20.000000     58.000000    7500.000000   
50%    1486.000000  5.354532e+08    50.000000     68.000000    9430.000000   
75%    2198.000000  9.071801e+08    70.000000     80.000000   11513.500000   
max    2930.000000  9.241520e+08   190.000000    313.000000  159000.000000   

       Overall Qual  Overall Cond   Year Built  Year Remod/Add  Mas Vnr Area  \
count   2051.000000   2051.000000  2051.000000     2051.000000   2029.000000   
mean       6.112140      5.562165  1971.708922     1984.190151     99.695909   
std        1.426271      1.104497    30.177889       21.036250    174.963129   
min        1.000000      1.000000  1872.000000     1950.000000      0.000000   
25%        5.000000      5.000000  1953.500000     1964.500000      0.000000   
50%        6.000000      5.000000  1974.000000     1993.000000      0.000000   
75%        7.000000      6.000000  2001.000000     2004.000000    161.000000   
max       10.000000      9.000000  2010.000000     2010.000000   1600.000000   

           ...        Wood Deck SF  Open Porch SF  Enclosed Porch  \
count      ...         2051.000000    2051.000000     2051.000000   
mean       ...           93.833740      47.556802       22.571916   
std        ...          128.549416      66.747241       59.845110   
min        ...            0.000000       0.000000        0.000000   
25%        ...            0.000000       0.000000        0.000000   
50%        ...            0.000000      27.000000        0.000000   
75%        ...          168.000000      70.000000        0.000000   
max        ...         1424.000000     547.000000      432.000000   

        3Ssn Porch  Screen Porch    Pool Area      Misc Val      Mo Sold  \
count  2051.000000   2051.000000  2051.000000   2051.000000  2051.000000   
mean      2.591419     16.511458     2.397855     51.574354     6.219893   
std      25.229615     57.374204    37.782570    573.393985     2.744736   
min       0.000000      0.000000     0.000000      0.000000     1.000000   
25%       0.000000      0.000000     0.000000      0.000000     4.000000   
50%       0.000000      0.000000     0.000000      0.000000     6.000000   
75%       0.000000      0.000000     0.000000      0.000000     8.000000   
max     508.000000    490.000000   800.000000  17000.000000    12.000000   

           Yr Sold      SalePrice  
count  2051.000000    2051.000000  
mean   2007.775719  181469.701609  
std       1.312014   79258.659352  
min    2006.000000   12789.000000  
25%    2007.000000  129825.000000  
50%    2008.000000  162500.000000  
75%    2009.000000  214000.000000  
max    2010.000000  611657.000000  

[8 rows x 39 columns]

##### cleaning:

#columns with the most missing values:

Alley (140)

Pool QC (9)

Fence (400)

Misc Features (65)

next worse column is Lot Frontage (1721)
- I'm going to make 2 models, one with lot frontage and missing ~300 entries, and one without lot frontage


In [6]:
ames.drop(labels=["Alley","Pool QC","Fence","Misc Feature"], axis=1, inplace=True)

In [7]:
ames_frtg = ames.dropna(axis=0)
ames_frtg.shape

(812, 77)

In [8]:
ames_nofrtg = ames.drop(labels=["Lot Frontage"], axis=1)
ames_nofrtg.dropna(inplace=True)
ames_nofrtg.shape

(1015, 76)

### (Basic) first attempt

In [9]:
# original feature selection from "eyeballing" the data
features = ["Lot Area", "1st Flr SF", "Overall Qual", "Overall Cond", "2nd Flr SF", "Pool Area"]
X = ames[features]
y = ames.SalePrice

In [10]:
lr = LinearRegression()

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X,y)

In [12]:
scores = cross_val_score(lr, X_train, y_train)
print(scores.mean())

0.7732688138054363


In [13]:
lr.fit(X,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [14]:
X_holdout = test_set[features]

In [15]:
y_hat = lr.predict(X_holdout)

In [16]:
my_ids = test_set["Id"]

In [17]:
sub = pd.DataFrame()

In [18]:
sub["Id"] = my_ids
sub["SalePrice"] = y_hat
sub.set_index(["Id"], inplace=True)

In [19]:
sub.head()

SalePrice
Id                 
2658  187245.387192
2718  194291.847231
2414  198605.570807
1989  117049.506597
625   181386.951645

In [20]:
sub.to_csv("./submission_2")

In [21]:
#ames["Misc Feature"].value_counts()

### Feature Engineering and Selection, Modeling

In [22]:
test_set.drop(labels=["Alley","Pool QC","Fence","Misc Feature"], axis=1, inplace=True)

In [23]:
test_set.dropna(inplace=True)

In [27]:
test_set.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
2   2414  528218130           60        RL          58.0     17104   Pave   
9   1568  914476010           20        RL          75.0     10125   Pave   
12  2584  535303050           20        RL          75.0     11310   Pave   
14  2673  903206160           70        RL          69.0     11737   Pave   
16  1297  902128150          190        RM          60.0     10800   Pave   

   Lot Shape Land Contour Utilities    ...     Wood Deck SF Open Porch SF  \
2        IR1          Lvl    AllPub    ...              100            24   
9        Reg          Lvl    AllPub    ...              238            83   
12       Reg          Lvl    AllPub    ...                0            30   
14       IR1          Bnk    AllPub    ...               27            36   
16       Reg          Lvl    AllPub    ...                0             0   

   Enclosed Porch 3Ssn Porch Screen Porch Pool Area Misc Val  Mo Sold  \
2               0          0            0         0        0        9   
9               0          0            0         0        0        2   
12              0          0            0         0        0        6   
14             42          0            0         0        0        5   
16            126          0            0         0        0        5   

    Yr Sold  Sale Type  
2      2006        New  
9      2008        WD   
12     2006        WD   
14     2006        WD   
16     2008        Con  

[5 rows x 76 columns]

In [32]:
# making sure that the test_set and the data_sets have the same number of dummy columns
test_set_lf = test_set
test_set_nlf = test_set
target = "SalePrice"
features_lf = [x for x in ames_frtg.columns if x != target]

# i was getting dummy columns in the 9000 range without making a separate list of all object columns
# courtesy of Mike T. for the help with normalizing the shape of the test and train data sets 

to_be_dummed = ["MS Zoning", "Street", "Lot Shape", "Land Contour", "Utilities", "Lot Config", "Land Slope",
                "Neighborhood", "Condition 1", "Condition 2", "Bldg Type", "House Style", "Roof Style", "Roof Matl",
                "Exterior 1st", "Exterior 2nd", "Mas Vnr Type", "Exter Qual", "Exter Cond", "Foundation", 
                "Bsmt Qual", "Bsmt Cond", "Bsmt Exposure", "BsmtFin Type 1", "BsmtFin Type 2", "Heating", "Heating QC", 
                "Central Air", "Electrical", "Kitchen Qual", "Functional" , "Fireplace Qu", "Garage Type", "Garage Finish",
                "Garage Qual", "Garage Cond", "Paved Drive","Sale Type"]

for i in to_be_dummed:
    testsetlf = list(test_set_lf[i].unique())
    amesfrtg = list(ames_frtg[i].unique())
    combinedlist = set(testsetlf+amesfrtg)
    test_set_lf[i] = pd.Categorical(test_set_lf[i], categories=combinedlist)
    ames_frtg[i] = pd.Categorical(ames_frtg[i], categories=combinedlist)

/Users/brianconnor/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


In [35]:
testsetlfdummied = pd.get_dummies(test_set_lf)
ames_frtg = pd.get_dummies(ames_frtg)

print(testsetlfdummied.shape)
print(ames_frtg.shape)

(344, 253)
(812, 254)


#### First, with frontage included

In [37]:
# feature interaction

poly_lf = PolynomialFeatures(include_bias=False)
target = "SalePrice"
features_lf = [x for x in ames_frtg.columns if x != target]
ames_frtg_poly = poly_lf.fit_transform(ames_frtg[features_lf])
print(ames_frtg_poly.shape)
# ames_frtg_poly is my interaction array of predictors, to be used as "X" in the future

ames_frtg_poly_df = pd.DataFrame(ames_frtg_poly, columns=poly_lf.get_feature_names(features_lf))

(812, 32384)


In [38]:
# split and standardize
y = ames_frtg.SalePrice
Xf_train, Xf_test, yf_train, yf_test = train_test_split(ames_frtg_poly_df, y)

ss_frtg = StandardScaler()
ss_frtg.fit(Xf_train)
Xfs_train = ss_frtg.transform(Xf_train)
Xfs_test = ss_frtg.transform(Xf_test)

In [39]:
#setting up lasso
lasso_frtg = LassoCV(max_iter=10000)
lasso_frtg.fit(Xfs_train, yf_train)

LassoCV(alphas=None, copy_X=True, cv=None, eps=0.001, fit_intercept=True,
    max_iter=10000, n_alphas=100, n_jobs=1, normalize=False,
    positive=False, precompute='auto', random_state=None,
    selection='cyclic', tol=0.0001, verbose=False)

In [40]:
print(lasso_frtg.alpha_)

3206.2712221681354


In [41]:
#scoring the lasso
frtg_score = cross_val_score(lasso_frtg, Xfs_train, yf_train)

In [42]:
frtg_score.mean()

0.7982078791644596

In [43]:
#scoring against the test set
lasso_frtg.score(Xfs_test, yf_test)

0.5586838819806191

In [ ]:
#in order to predict yhat from the holdout set, I need to transform the data in the holdout the same way I transformed
#the given data. I dummied it before setting up the Lasso, so now I think I just need to poly and standardize

In [46]:
testsetlfdummied.shape

(344, 253)

In [47]:
test_set_lf_poly = poly_lf.transform(testsetlfdummied)

In [48]:
test_set_lf_ready = ss_frtg.transform(test_set_lf_poly)

In [67]:
yhat_frtg = lasso_frtg.predict(test_set_lf_ready)


(344, 253)

In [68]:
print(yhat_frtg.shape)
print(testsetlfdummied.shape)

(344,)
(344, 253)


In [69]:
frtg_id = testsetlfdummied["Id"]
sub2 = pd.DataFrame()
sub2["Id"] = frtg_id
sub2["SalePrice"] = yhat_frtg
sub2.set_index(["Id"], inplace=True)

In [73]:
sub2.shape

(344, 1)

In [72]:
sub.to_csv("./frtg_sub")

## This submission doesn't work because you can't remove the NaN entries from the test set

#### Now with no frontage

In [ ]:
#ames_nofrtg

#dummies
ames_nofrtg = pd.get_dummies(ames_nofrtg)

In [ ]:
#feature interaction
poly_nolf = PolynomialFeatures(include_bias=False)
target = "SalePrice"
features_nolf = [x for x in ames_nofrtg.columns if x != target]
ames_nofrtg_poly = poly_nolf.fit_transform(ames_nofrtg[features_nolf])
ames_nofrtg_poly.shape

# ames_frtg_poly is my interaction array of predictors, to be used as "X" in the future

ames_nofrtg_poly_df = pd.DataFrame(ames_nofrtg_poly, columns=poly_nolf.get_feature_names(features_nolf))

In [ ]:
# split and standardize
ynf = ames_nofrtg.SalePrice
Xnf_train, Xnf_test, ynf_train, ynf_test = train_test_split(ames_nofrtg_poly_df, ynf)

ss_nofrtg = StandardScaler()
ss_nofrtg.fit(Xnf_train)
Xnfs_train = ss_nofrtg.transform(Xnf_train)
Xnfs_test = ss_nofrtg.transform(Xnf_test)

In [ ]:
#setting up lasso
lasso_nofrtg = LassoCV(max_iter=10000)
lasso_nofrtg.fit(Xnfs_train, ynf_train)

In [ ]:
print(lasso_nofrtg.alpha_)

In [ ]:
#scoring lasso
nofrtg_score = cross_val_score(lasso_nofrtg, Xnfs_train, ynf_train)

In [ ]:
nofrtg_score.mean()

In [ ]:
lasso_nofrtg.score(Xnfs_test, ynf_test)